### ISSUE TITLE
More flexible signing interface

### ISSUE THREAD

**Issue by radeksimko** [goreleaser/goreleaser#1411]
I'm attempting to sign MacOS & Windows binaries by sending them over to a separate (centralised) CI pipeline for signing and downloading signed ones back from there.

I admit this may not be a common scenario, but I thought I'd open a conversation here to see if you were interested in supporting it somehow, also because I feel that some of the problems described here apply to people who sign locally with `gon` anyway.

--- 

There is a few problems I ran into while attempting to implement this, all related to the current signing interface (as represented by `signs` section of the config):

1. It is assumed that signature will be stored in a separate file such as `<artifact>.sig` - this is common for GPG-based signing, but not suitable for MacOS or Windows signing, where the binary is basically modified in place. I suppose this can be overwritten via `signature: "${artifact}"` but it seems hacky, or non-intuitive at the least as GoReleaser doesn't seem to have a notion of `SignedBinary` or `SignedArchive` artifact type, just `Signature`.
2. While there is filtering of artifacts there is no way to filter _just archives_ (to avoid notarizing checksum files) - the only options available are `none`, `all` and `checksums`
3. It is assumed that signing takes place _after_ archival, so the only artifacts available are actually archives, rather than binaries
4. It is assumed that checksum calculation takes place before signing, so checksums won't match for archives which have binaries replaced with signed ones.

**Comment by radeksimko** [comment]
After thinking about this a bit more I feel that instead of trying to retrofit the above signing process into the current signing interface it may be easier to implement it as post build hooks - I left some ideas here:
https://github.com/goreleaser/goreleaser/issues/1328

**Comment by caarlos0** [comment]
fixed by #1414 then?

**Comment by radeksimko** [comment]
Yes, but I feel [this section](https://goreleaser.com/customization/#signing-with-gon) should probably recommend the build hooks instead of the first-class signing interface, which seems to be mainly aimed at signing archives? Or perhaps it should even say that signing binaries in general (where `gon` is one of the ways to do it) should happen in build hooks and `signs` should only be used for signing archives.

I suspect that redesigning the `signs` to cater the signing needs described above would be quite a lot of work? Not sure what your plans were there 🤔 (whether you intended to support code-signing) - so in theory `signs` could even be renamed to `archive_signs` in v1? 🤷‍♂️ 

Alternatively - do you think there's any point in introducing some sort of first-class binary signing section? I'm leaning more towards no, but interested to hear your opinion.

--- 

TL;DR this is now technically possible, but _how_ to do it right may not be obvious for users from naming conventions in config nor docs.

**Comment by caarlos0** [comment]
> Yes, but I feel [this section](https://goreleaser.com/customization/#signing-with-gon) should probably recommend the build hooks instead of the first-class signing interface, which seems to be mainly aimed at signing archives? Or perhaps it should even say that signing binaries in general (where `gon` is one of the ways to do it) should happen in build hooks and `signs` should only be used for signing archives.

makes sense! do you feel like opening a pr improving it?

> I suspect that redesigning the `signs` to cater the signing needs described above would be quite a lot of work? Not sure what your plans were there 🤔 (whether you intended to support code-signing) - so in theory `signs` could even be renamed to `archive_signs` in v1? 🤷‍♂️

yeah, or maybe we can use the same "hooks" thing from builds, but on archives instead 🤔

> Alternatively - do you think there's any point in introducing some sort of first-class binary signing section (alongside the existing one)? I'm leaning more towards no, but interested to hear your opinion.

yeah, maybe it makes sense as well... but the hooks thing make it simpler I think 🤔

**Comment by radeksimko** [comment]
> do you feel like opening a pr improving it?

Yep, I can craft something.

> or maybe we can use the same "hooks" thing from builds, but on archives instead

That's a good point - hooks as a concept are certainly more versatile. I like that idea, although I have never used the existing signing interface, so I don't know how much pain would it cause to deprecate it in favour of more generic hooks.

**Comment by caarlos0** [comment]
> That's a good point - hooks as a concept are certainly more versatile. I like that idea, although I have never used the existing signing interface, so I don't know how much pain would it cause to deprecate it in favour of more generic hooks.

usually, I add a deprecation notice with "how to migrate", wait ~6 months, remove the deprecated thing 🤔

https://goreleaser.com/deprecations/
